apiclient/harvester_api/managers.py

```diff
@@ -105,11 +105,11 @@ def delete(self, name, *, raw=False):
     def get_metrics(self, name="", *, raw=False):
         return self._get(self.METRIC_fmt.format(uid=name), raw=raw)
 
-    def maintenance_mode(self, name, enable=True):
+    def maintenance_mode(self, name, enable=True, force=False):
         action = "enable" if enable else "disable"
         params = dict(action=f"{action}MaintenanceMode")
-        self._create(self.PATH_fmt.format(uid=name), params=params)
-        return self.get(name)
+        payload = dict(force=str(force).lower())
```

Comment:
Do we need to do a `.lower()` there?
It doesn't seem like it matters to the API backend.
With for instance:
```
curl --location --request POST 'https://192.168.1.186/v1/harvester/nodes/v11-head-top?action=enableMaintenanceMode' \
--header 'Authorization: Bearer token-wz48k:4c8rwf6lfqcvb2q6c688vj5s6ltvqx2kclkfs9cvmcg9gfrg2h6vz2' \
--header 'Content-Type: application/json' \
--data-raw '{
    "force": "TrUe"
}'
```
That returns a 204 and does put the node into maintenance mode, forcefully as far as I can tell.
I also wasn't able to find this documented in the OpenAPI swagger.json spec, I wonder if this could be added to show the users that this can be an API action:
https://github.com/harvester/harvester/blob/master/api/openapi-spec/swagger.json

Reply:
The parameter be introduced in https://github.com/harvester/harvester/issues/2723, and you can see we will use `true` to identify whether it is **forced** in the switch case.[^1]

The PR didn't update swagger API spec, so it isn't listed in the API.

[^1]: https://github.com/harvester/harvester/blob/9f440f9fc299bb1fe5404aac88ee5ac57ca1522d/pkg/api/node/formatter.go#L98

Reply:
@lanfon72  - Thanks for linking that :+1: !

I guess, I'm surprised by the request of:
```
curl --location --request POST 'https://192.168.1.186/v1/harvester/nodes/v11-head-top?action=enableMaintenanceMode' \
--header 'Authorization: Bearer token-wz48k:4c8rwf6lfqcvb2q6c688vj5s6ltvqx2kclkfs9cvmcg9gfrg2h6vz2' \
--header 'Content-Type: application/json' \
--data-raw '{
    "force": "TrUe"
}'
```
Getting a `204` back.
This sounds like an API question I suppose, since the server I'm sure understands the request, but the JSON payload is incorrect. 
On the backend, it *must* have a case-sensitive nature, as in:
https://github.com/harvester/harvester/blob/master/pkg/api/node/formatter.go#L126-L143
( enabling maintenance mode )
in the source it relies on an exact match of "true" ( with the casing being respected )
Same with:
https://github.com/harvester/harvester/blob/master/pkg/api/node/formatter.go#L92-L101

I'm just surprised because I think the end user would also like to know that:
```
{
   "force": "TrUe"
}
```
the value of `TrUe` is invalid and shouldn't be sending back a `204` perhaps send back instead a `422`.

Also curious as to why the design was to send a `string` of "true" vs something like:
```json
{
    "force": true
}
```
boolean values are valid JSON... :thinking: 

I'm just confused as to how the end user would know to pass a "case sensitive" value when the feedback from the api is a `204`.  Like how they would know "they" made a mistake in the payload value with the `204`.  Like the node will be put into maintenance mode regardless but if they wanted to force the node into maintenance mode and shipped the wrong JSON payload, they wouldn't know from the response that it was inadequate.

Reply:
> the value of `TrUe` is invalid and shouldn't be sending back a `204` perhaps send back instead a `422`.

It returns `204` means the `maintenance mode` been applied, the `force` option is the additional way to **forced** drain nodes, if it is not matched, maintenance mode will still applied without **forced** drain, so it should not return `422` but `204`.

> I'm just confused as to how the end user would know to pass a "case sensitive" value when the feedback from the api is a `204`. 

We don't have any API library yet, and API document not include this, so end user will only able to use the function by dashboard or the `harvester_api` (apiclient) in this repo.

To use boolean or not, It is design decision, I thought you can discuss with @ibrokethecloud about the details.

Reply:
that's a great point :smile: 
yeah, since it would still put it into maintenance mode regardless of the `force` option being passed in `422` doesn't make sense.
I looked at those code snippets again this morning and it makes more sense :+1: :smile: 

I guess I'm still curious surrounding the case sensitivity but if a user is typically going to only ever trigger this via the dashboard then I suppose it doesn't matter that much :smile: